### PR TITLE
Feat: IBM watsonx.ai Chat integration

### PR DESCRIPTION
**Pull request by Wojciech-Rebisz**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Summary:
The purpose of these changes is to add Chat API support to the IBM watsonx.ai integration. Such will accept the `tool_choice` and `tool_choice_option` options. Methods introduced/changed: 
 - `chat() `
 - `stream_chat()`

FYI @Mateusz-Switala @LukaszCmielowski
## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [x] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Your pull-request will likely not be merged unless it is covered by some form of impactful unit testing.

- [x] I believe this change is already covered by existing unit tests
- [x] I added mocks to use a response that reflects the actual content (in chat & chat_stream methods related tests)

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py [hunk 1]
```diff
@@ -58,10 +69,10 @@ class WatsonxLLM(CustomLLM):
     """
 
     model_id: Optional[str] = Field(
-        default=None, description="Type of model to use.", frozen=True
+        default=None, description="Type of model to use.", allow_mutation=False
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py:72]: allow_mutations is the pydantic v1 way no? For pydantic v2, it should be frozen=True

https://docs.pydantic.dev/latest/concepts/fields/#immutability

File: llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py [hunk 2]
```diff
@@ -245,7 +256,8 @@ def __init__(
         self._model_info = None
         self._deployment_info = None
 
-    model_config = ConfigDict(protected_namespaces=(), validate_assignment=True)
+    class Config:
+        validate_assignment = True
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py:260]: this is the pydantic v1 way of setting a config. Can we revert back to `model_config = ConfigDict(protected_namespaces=(), validate_assignment=True)` ?

https://docs.pydantic.dev/latest/concepts/config/

File: llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py [hunk 3]
```diff
@@ -335,6 +365,12 @@ def complete(
             raw=response,
         )
 
+    @llm_completion_callback()
+    async def acomplete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> CompletionResponse:
+        return self.complete(prompt, formatted=formatted, **kwargs)
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py:372]: there's no async client or async client methods?

**Review by logan-markewich — COMMENTED** [review]
This looks fine to me, just a note about some pydantic v1 vs v2 changes

File: llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py (hunk 1 above)
Comment (Wojciech-Rebisz) [reply llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py]: Fixed

File: llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py (hunk 2 above)
Comment (Wojciech-Rebisz) [reply llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py]: Reverted

File: llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py (hunk 3 above)
Comment (Wojciech-Rebisz) [reply llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py:372]: Part of them are still under construction in the source package, so they will be delivered soon (as the next step).

Comment (logan-markewich) [reply llama-index-integrations/llms/llama-index-llms-ibm/llama_index/llms/ibm/base.py:372]: Great!

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**